src/Processors/Transforms/buildPushingToViewsChain.cpp

```diff
@@ -270,11 +270,14 @@ Chain buildPushingToViewsChain(
         ASTPtr query;
         Chain out;
 
+        /// NOTE: ThreadGroupStatus always should have context attached,
+        /// otherwise entry to the system.query_views_log will not be added
+        /// (see ThreadStatus::logToQueryViewsLog())
         ThreadGroupStatusPtr running_group;
-        if (current_thread && current_thread->getThreadGroup())
+        if (current_thread)
             running_group = current_thread->getThreadGroup();
-        else
-            running_group = std::make_shared<ThreadGroupStatus>();
+        if (!running_group)
+            running_group = std::make_shared<ThreadGroupStatus>(context);
```

Comment:
Can you help with the question? When there is a current_thread->getThreadGroup and when there is no group in current thread?

One more question, in case then current_thread do not have TG, why is it ring to make a TG for each view insted make one for all?

Reply:
Oh I see. The is recursion call buildPushingToViewsChain. And that call appears 6 times in the code. Actually, as for me, it not so much to touch. An only one of them is trigger: src/Interpreters/InterpreterInsertQuery.cpp:276.

In order to achieve one TG for all, we could either pass it through the function args buildPushingToViewsChain.

Or:

Move the block
```
ThreadGroupStatusPtr running_group;
        if (current_thread)
            running_group = current_thread->getThreadGroup();
        if (!running_group)
            running_group = std::make_shared<ThreadGroupStatus>(context);
```
out of the for loop. Then, when we had to make new TG, bind current thread with this new TG with ThreadGroupSwitcher for the scope of buildPushingToViewsChain. That makes following recursive calls to see the new TG. (Or rewrite it in other similat way: if no current_thread->getThreadGroup(), the create it add attach temporarily)

Actually, all this things about finding/creating/attaching TG to the thread should  be done between (or inside) InterpreterInsertQuery::buildChainImpl and buildPushingToViewsChain. In that way it looks more simple, without recursion context in the mind.

How do you see this? I do not insist for such approach. You are solving specific problem and the way is OK. 
However one TG for MVs allows to collect all the memory accounting and profiles counters in one place: that TG. In case if that had been done in scope if query, all that information would be binded to the query. Your solution already do that binding, that is cool.

Reply:
One thing: no need to use ThreadGroupSwitcher as long as we bind new thread group to the thread without thread group. That binding could be permanent.

Reply:
>Then, when we had to make new TG, bind current thread with this new TG with ThreadGroupSwitcher for the scope of buildPushingToViewsChain. That makes following recursive calls to see the new TG. (Or rewrite it in other similat way: if no current_thread->getThreadGroup(), the create it add attach temporarily)

Well, it will work, but this implicit usage looks not clean.

>Actually, all this things about finding/creating/attaching TG to the thread should be done between (or inside) InterpreterInsertQuery::buildChainImpl and buildPushingToViewsChain. In that way it looks more simple, without recursion context in the mind.

The reason I kept it in `buildPushingToViewsChain`, because it play low level games to manipulate the `ThreadStatus`, and this thing cannot moved to the `InterpreterInsertQuery`.

And this place will be moved to the `InterpreterInsertQuery`, then there will be two different places that will do this low level stuff, at first I thought that this is a bad idea, but while writing this paragraph, I slightly changed my mind, so I guess it is OK.

>One thing: no need to use ThreadGroupSwitcher as long as we bind new thread group to the thread without thread group. That binding could be permanent.

Yes, make sense, missed the patches where `MemoryTrackerThreadSwitcher` had been renamed to `ThreadGroupSwitcher`.

But I found it odd that it is located in the `MergeTree` code, so what I'm going to do is:
- move `ThreadGroupSwitcher` into a separate file
- rename `ThreadGroupStatus` to `ThreadGroup`
- fix this place with multiple TG attached here
But I think it is better to do this on top of this  PR in a separate one, @CheSema objections?

Reply:
Sounds good!
ThreadGroupSwitcher deserve better place indeed.
However here is it has no use since we create and bind designated TG only if there is no query scope TG. That binding could be permanent without ThreadGroupSwitcher. The other case: if there is a query scope TG, we attach it to all MV's thread statuses, in other words we reuse it.

Reply:
>That binding could be permanent without ThreadGroupSwitcher

Yes, indeed.

So here it is - #48543